pkg/ingester/incremental_transfer.go

```diff
@@ -0,0 +1,260 @@
+package ingester
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/ingester/client"
+	"github.com/cortexproject/cortex/pkg/ring"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/weaveworks/common/user"
+	"golang.org/x/net/context"
+)
+
+var (
+	blockedRanges = prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "cortex_ingester_blocked_ranges",
+		Help: "The current number of ranges that will not accept writes by this ingester.",
+	})
+)
+
+func init() {
+	prometheus.MustRegister(blockedRanges)
+}
+
+// TransferChunksSubset accepts chunks from a client and moves them into the local Ingester.
+func (i *Ingester) TransferChunksSubset(stream client.Ingester_TransferChunksSubsetServer) error {
+	i.userStatesMtx.Lock()
+	defer i.userStatesMtx.Unlock()
+
+	fromIngesterID, seriesReceived, err := i.acceptChunksFromStream(acceptChunksOptions{
+		States: i.userStates,
+		Stream: stream,
+	})
+	if err != nil {
+		return err
+	}
+
+	if err := stream.SendAndClose(&client.TransferChunksResponse{}); err != nil {
+		level.Error(util.Logger).Log("msg", "Error closing SendChunks stream", "from_ingester", fromIngesterID, "err", err)
+		return err
+	}
+
+	// Target ingester may not have had any streams to send.
+	if seriesReceived == 0 {
```

Comment:
This `if seriesReceived == 0` check seems unnecssary as you're returning nil regardless.